jetty-client/src/main/scala/org/http4s/client/jetty/JettyClient.scala

```diff
@@ -0,0 +1,131 @@
+package org.http4s
+package client
+package jetty
+
+import cats.data.Kleisli
+import cats.effect._
+import cats.effect.implicits._
+import cats.implicits.{catsSyntaxEither => _, _}
+import fs2._
+import fs2.Stream._
+import java.nio.ByteBuffer
+import org.eclipse.jetty.client.HttpClient
+import org.eclipse.jetty.client.api.{Result, Request => JettyRequest, Response => JettyResponse}
+import org.eclipse.jetty.client.util.DeferredContentProvider
+import org.eclipse.jetty.http.HttpFields
+import org.log4s.getLogger
+import scala.collection.JavaConverters._
+import scala.concurrent.ExecutionContext
+import scala.util.control.NonFatal
+
+object JettyClient {
+
+  private val logger = getLogger
+
+  def apply[F[_]](client: HttpClient = new HttpClient())(
+      implicit F: ConcurrentEffect[F],
+      ec: ExecutionContext): F[Client[F]] =
+    F.pure(client)
+      .flatTap(client => F.delay { client.start() })
+      .map(client => {
+        Client(
+          Kleisli { req =>
+            F.async[DisposableResponse[F]] { cb =>
+              toRequest(client, req).send(responseHandler(cb))
+            }
+          },
+          F.delay({
+            try {
+              client.stop()
+            } catch {
+              case NonFatal(t) => logger.warn(t)("Unable to shut down Jetty client")
+            }
+          })
+        )
+      })
+
+  def stream[F[_]](client: HttpClient = new HttpClient())(
+      implicit F: ConcurrentEffect[F],
+      ec: ExecutionContext): Stream[F, Client[F]] =
+    Stream.bracket(apply(client))(_.shutdown)
+
+  private def toRequest[F[_]](client: HttpClient, request: Request[F])(
+      implicit F: Effect[F],
+      ec: ExecutionContext): JettyRequest = {
+    val jReq = client.newRequest(request.uri.toString).method(request.method.name)
+
+    for (h <- request.headers)
+      jReq.header(h.name.toString, h.value)
+
+    jReq.content(requestContentProvider(request))
+  }
+
+  private def requestContentProvider[F[_]](
+      request: Request[F])(implicit F: Effect[F], ec: ExecutionContext) = {
+    val dcp = new DeferredContentProvider()
+
+    ec.execute(new Runnable {
+      override def run(): Unit =
+        request.body.chunks
+          .map(_.toByteBuffer)
+          .to(Sink((b: ByteBuffer) => F.delay { dcp.offer(b); () }))
```

Comment:
Oh, the Jetty Chunk's `Callback` is invoked when the content is read.  I think we probably want to use `Chunk` instead of `ByteBuffer` here, because with `Chunk`, you could implement backpressure.  As it stands, we're just going to fill up the `dcp` without any regard to how fast the server consumes it.

Reply:
true, right now the dcp is gonna be loaded regardless of any circumstances. I'm not entirely sure how to implement the back pressure mechanism but I'll try to figure out something, I guess that each `dcp.offer` should somehow wait on the previous callback to be called with `.succeed()` or `.failed(Throwable t)` and decide either to continue or to stop processing, right? 

Btw, I'm wondering if by using the [InputStreamContentProvider](https://github.com/eclipse/jetty.project/blob/jetty-9.4.x/jetty-client/src/main/java/org/eclipse/jetty/client/util/InputStreamContentProvider.java#L54) we would get this semantic for free (but it would obviously block on `InputStream.read`)

Reply:
The [PublisherContentProvider](https://github.com/jetty-project/jetty-reactive-httpclient/blob/9bf80e8c25ad10cdaa3aeb3d5af0b51a0ffe3ec3/src/main/java/org/eclipse/jetty/reactive/client/internal/PublisherContentProvider.java#L67) isn't _exactly_ the same thing, but might be an inspiration.

What about a semaphore with 1 permit?  We take a permit to emit a chunk, and return a permit on each callback?